File: SeaBattle/SeaBattle.CSharp/BoardCell.cs
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace SeatBattle.CSharp
{
    [DebuggerDisplay("({X},{Y}) {_state}")]
    public class BoardCell : Label
    {
        private static readonly Color DefaultBorderColor = Color.FromArgb(214,214,214);
        private static readonly Color DefaultBackgroundColor = Color.FromArgb(222,222,222);
        private static readonly Color DragOverBackgroundColor = Color.FromArgb(255,174,0);
        private static readonly Color DragOverInvalidBackgroundColor = Color.FromArgb(222,0,0);
        private static readonly Color ShipColor = Color.FromArgb(65,133,243);
        private static readonly Color ShipDrownedColor = Color.FromArgb(222, 0, 0);

        private const char ShipHitChar = (char)0x72;
        private const char MissedHitChar = (char)0x3D;



        private BoardCellState _state;

        public BoardCell(int x, int y)
        {
            X = x;
            Y = y;
            base.AutoSize = false;
            base.TextAlign = ContentAlignment.MiddleCenter;
            base.Font = new Font("Webdings", 10);
            base.AllowDrop = true;
        }

        public BoardCellState State
        {
            get
            {
                return _state;
            }
            set
            {
                _state = value;
                OnCellStateChenged();
            }
        }

        private void OnCellStateChenged()
        {
            SuspendLayout();
            switch (_state)
            {
                case BoardCellState.Normal:
                    Text = string.Empty;
                    BackColor = DefaultBackgroundColor;
                    break;
                case BoardCellState.MissedShot:
                    Text = MissedHitChar.ToString();
                    BackColor = DefaultBackgroundColor;
                    break;
                case BoardCellState.Ship:
                    Text = string.Empty;
                    BackColor = ShipColor;
                    break;
                case BoardCellState.ShotShip:
                    Text = ShipHitChar.ToString();
                    BackColor = ShipColor;
                    break;
                case BoardCellState.ShipDrag:
                    BackColor = DragOverBackgroundColor;
                    Text = string.Empty;
                    break;
                case BoardCellState.ShipDragInvalid:
                    BackColor = DragOverInvalidBackgroundColor;
                    Text = string.Empty;
                    break;
                case BoardCellState.ShowDrowned:
                    BackColor = ShipDrownedColor;
                    Text = ShipHitChar.ToString();
                    break;
            }
            Invalidate();
            ResumeLayout();
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            base.OnPaint(e);

            using (var pen = new Pen(DefaultBorderColor))
            {
                pen.Alignment = PenAlignment.Inset;
                pen.DashStyle = DashStyle.Solid;

                var rect = ClientRectangle;
                rect.Height -= 1;
                rect.Width -= 1;

                e.Graphics.DrawRectangle(pen, rect);
            }
        }

        public int X { get; private set; }
        public int Y { get; private set; }
    }
}

File: SeaBattle/SeaBattle.Cpp/BoardCell.cpp
#include "stdafx.h"
#include "BoardCell.h"

BoardCell::BoardCell(int x, int y)
{
	X = x;
	Y = y;
	AutoSize = false;
	TextAlign = ContentAlignment::MiddleCenter;
	BackColor = Color::LightBlue;
	Font = gcnew System::Drawing::Font("Webdings", 10);
	AllowDrop = true;
}

void BoardCell::OnCellStateChenged()
{
	SuspendLayout();
	switch (_state)
	{
		case BoardCellState::Normal:
			Text = String::Empty;
			BackColor = DefaultBackgroundColor;
			break;
		case BoardCellState::MissedShot:
			Text = MissedHitChar->ToString();
			BackColor = DefaultBackgroundColor;
			break;
		case BoardCellState::Ship:
			Text = String::Empty;
			BackColor = ShipColor;
			break;
		case BoardCellState::ShotShip:
			Text = ShipHitChar->ToString();
			BackColor = ShipColor;
			break;
		case BoardCellState::ShipDrag:
			BackColor = DragOverBackgroundColor;
			Text = String::Empty;
			break;
		case BoardCellState::ShipDragInvalid:
			BackColor = DragOverInvalidBackgroundColor;
			Text = String::Empty;
			break;
		case BoardCellState::ShowDrowned:
			BackColor = ShipDrownedColor;
			Text = ShipHitChar->ToString();
			break;
	}
	Invalidate();
	ResumeLayout();
}

void BoardCell::OnPaint(PaintEventArgs ^e)
{
	Label::OnPaint(e);

	Pen^ pen = gcnew Pen(DefaultBorderColor);
	pen->Alignment = PenAlignment::Inset;
	pen->DashStyle = DashStyle::Solid;

	Rectangle rect = ClientRectangle;
	rect.Height -= 1;
	rect.Width -= 1;

	e->Graphics->DrawRectangle(pen, rect);

    delete pen;
}

BoardCellState BoardCell::State::get()
{
	return _state;
}

void BoardCell::State::set(BoardCellState value)
{
	_state = value;
	OnCellStateChenged();
}